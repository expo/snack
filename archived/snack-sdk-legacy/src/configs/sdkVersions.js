/* @flow */

import semver from 'semver';

// minimum SDK versions that support snack features
const minFeatureVersion = {
  MULTIPLE_FILES: '21.0.0',
  PROJECT_DEPENDENCIES: '25.0.0',
  TYPESCRIPT: '31.0.0',
  UNIMODULE_IMPORTS: '33.0.0',
  POSTMESSAGE_TRANSPORT: '35.0.0',
  VERSIONED_SNACKAGER: '37.0.0',
};

// special casing of features that have been backported to particular SDK versions
export const versions = {
  '37.0.0': [],
  '38.0.0': [],
  '39.0.0': [],
  '40.0.0': [],
};

export const defaultSDKVersion = '40.0.0';

export type Feature = $Keys<typeof minFeatureVersion>;
export type SDKVersion = $Keys<typeof versions>;

export const sdkSupportsFeature = (sdkVersion: SDKVersion, feature: Feature) => {
  const versionsIncludeFeature = versions[sdkVersion]
    ? versions[sdkVersion].includes(feature)
    : false;
  const result = semver.gte(sdkVersion, minFeatureVersion[feature]) || versionsIncludeFeature;
  return result;
};
