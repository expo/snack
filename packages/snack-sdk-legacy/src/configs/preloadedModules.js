export const aliases = {
  // @expo/snack-static/react-native-logo.png is an alias that is
  // implemented in the runtime, and uses by the react-native image docs
  '@expo/snack-static': '0.0.1',
  '@expo/snack-static/react-native-logo.png': '0.0.1',
};

export const dependencies = {
  '37.0.0': {
    ...aliases,
    '@expo/vector-icons': '10.0.0',
    '@unimodules/core': '5.1.0',
    '@unimodules/react-native-adapter': '5.1.1',
    expo: '37.0.0',
    'expo-asset': '8.1.1',
    'expo-barcode-scanner': '8.1.0',
    'expo-camera': '8.2.0',
    'expo-constants': '9.0.0',
    'expo-file-system': '8.1.0',
    'expo-font': '8.1.0',
    'expo-gl': '8.1.0',
    'expo-image-picker': '8.1.0',
    'expo-linear-gradient': '8.1.0',
    react: '16.9.0',
    'react-native': '0.61.4',
    'prop-types': '15.7.2',
    'react-dom': '16.9.0',
    'react-native-web': '0.11.7',
    'react-native-gesture-handler': '1.6.0',
    'unimodules-permissions-interface': '5.1.0',
  },
  // The outcommented dependencies are also preloaded in the runtime, however we don't mark them
  // as such, so that the snack-sdk will allow adding them to the package.json dependencies.
  // When any of these dependencies is sent to the runtime, the version is ignored and the
  // runtime uses the version that was bundled into the runtime instead.
  '38.0.0': {
    ...aliases,
    expo: '38.0.1',
    react: '16.11.0',
    'react-native': '0.62.2',
    'react-dom': '16.11.0',
    'react-native-web': '0.11.7',
    // Packages that require special initialisation (see Modules.tsx)
    'expo-asset': '8.1.7',
    'react-native-gesture-handler': '1.6.0',
    //'react-native-vector-icons': '10.0.0',
    '@expo/vector-icons': '10.0.0',
    // Packages that are used internally by the runtime
    //'expo-analytics-amplitude': '8.2.1',
    'expo-barcode-scanner': '8.2.1',
    'expo-constants': '9.1.1',
    'expo-file-system': '9.0.1',
    //'expo-keep-awake': '8.2.1',
    'react-native-view-shot': '3.1.2',
    // Packages that are depended on by one of the packages above
    'expo-font': '8.2.1', // used by "expo"
    //'expo-linking': '1.0.3', // used by "expo"
    //'expo-permissions': '9.0.1', // used by "expo"
    //'expo-splash-screen': '0.3.1', // used by "expo"
    'prop-types': '15.7.2', // *
    '@unimodules/core': '5.3.0', // used by "expo"
    '@unimodules/react-native-adapter': '5.4.0', // used by "expo"
    'unimodules-permissions-interface': '5.2.1', // used by "expo"
    // Packages that are not really required
    // TODO: THESE SHOULD BE REMOVED ONCE THE SNACK-SDK HANDLES ADDING/REMOVING/UPDATING DEPENDENCIES BETTER
    'expo-camera': '8.3.1',
    'expo-gl': '8.3.1',
    'expo-image-picker': '8.3.0',
    'expo-linear-gradient': '8.2.1',
  },
  // The outcommented dependencies are also preloaded in the runtime, however we don't mark them
  // as such, so that the snack-sdk will allow adding them to the package.json dependencies.
  // When any of these dependencies is sent to the runtime, the version is ignored and the
  // runtime uses the version that was bundled into the runtime instead.
  '39.0.0': {
    ...aliases,
    expo: '39.0.2',
    react: '16.13.1',
    'react-native': '0.63.2',
    'react-dom': '16.13.1',
    'react-native-web': '0.13.12',
    // Packages that require special initialisation (see Modules.tsx)
    'expo-asset': '8.2.0',
    'react-native-gesture-handler': '1.7.0',
    //'react-native-vector-icons': '10.0.0',
    '@expo/vector-icons': '10.0.0',
    // Packages that are used internally by the runtime
    'expo-barcode-scanner': '9.0.0',
    'expo-constants': '9.2.0',
    'expo-file-system': '9.2.0',
    'react-native-view-shot': '3.1.2',
    // Packages that are depended on by one of the packages above
    'expo-font': '8.3.0', // used by "expo"
    'prop-types': '15.7.2', // *
    '@unimodules/core': '5.5.0', // used by "expo"
    '@unimodules/react-native-adapter': '5.6.0', // used by "expo"
    'unimodules-permissions-interface': '5.3.0', // used by "expo"
    // Packages that are not really required
    // TODO: THESE SHOULD BE REMOVED ONCE THE SNACK-SDK HANDLES ADDING/REMOVING/UPDATING DEPENDENCIES BETTER
    'expo-camera': '9.0.0',
    'expo-gl': '9.1.1',
    'expo-image-picker': '9.1.1',
    'expo-linear-gradient': '8.3.0',
  },
  '40.0.0': {
    ...aliases,
    expo: '40.0.0',
    react: '16.13.1',
    'react-native': '0.63.2',
    'react-dom': '16.13.1',
    'react-native-web': '0.13.12',

    // Packages that require special initialisation (see Modules.tsx)
    'expo-asset': '8.2.1',
    'expo-font': '8.4.0',
    'react-native-gesture-handler': '1.8.0',
    'react-native-safe-area-context': '3.1.9',
    //'react-native-vector-icons': '10.0.0',
    '@expo/vector-icons': '12.0.0',

    // Packages that are used internally by the runtime
    'expo-constants': '9.3.3',
    'expo-file-system': '9.3.0',
    'expo-permissions': '10.0.0',
    'expo-updates': '0.4.1',
    '@react-native-community/async-storage': '1.12.0',

    // Common packages that are included for easy of use
    'prop-types': '15.7.2',
    '@unimodules/core': '6.0.0',
    '@unimodules/react-native-adapter': '5.7.0',
    'unimodules-permissions-interface': '5.4.0',
  },
};
