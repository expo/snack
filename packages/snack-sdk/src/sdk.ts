import semver from 'semver';

import sdks from './sdks';
import features from './sdks/features';
import { SDKVersion, SDKFeature, SnackDependencies, SnackDependencyVersions } from './types';

/**
 * Checks whether a specific module/dependency is preloaded for the given SDK version.
 */
export function isModulePreloaded(
  name: string,
  sdkVersion: SDKVersion,
  coreModulesOnly?: boolean
): boolean {
  const sdk = sdks[sdkVersion];
  if (!sdk) return false;
  return !!(sdk.coreModules[name] || (!coreModulesOnly && sdk.bundledModules[name]));
}

/**
 * Returns the list of pre-loaded modules for the given SDK version.
 */
export function getPreloadedModules(
  sdkVersion: SDKVersion,
  coreModulesOnly?: boolean
): { [name: string]: string } {
  const sdk = sdks[sdkVersion];
  if (!sdk) return {};
  return coreModulesOnly
    ? sdk.coreModules
    : {
        ...sdk.coreModules,
        ...sdk.bundledModules,
      };
}

export function validateSDKVersion(sdkVersion: SDKVersion): SDKVersion {
  if (Object.keys(sdks).indexOf(sdkVersion) < 0) {
    throw new Error(
      `Invalid SDKVersion, the following versions are supported: ${Object.keys(sdks)}`
    );
  }
  return sdkVersion;
}

/**
 * Checks whether the module is deprecated for the given sdk-version and returns
 * the alternative module or error description instead. If the module is not deprecated
 * `undefined` is returned.
 */
export function getDeprecatedModule(name: string, sdkVersion: SDKVersion): string | undefined {
  return sdks[sdkVersion]?.deprecatedModules?.[name];
}

/**
 * Returns the list of supported SDK versions.
 */
export function getSupportedSDKVersions(): SDKVersion[] {
  return Object.keys(sdks) as SDKVersion[];
}

/**
 * Verifies whether a string is a valid semver.
 */
export function isValidSemver(version: string): boolean {
  return version === 'latest' || !!semver.validRange(version);
}

/**
 * Checks whether a feature is supported by the given SDK version.
 */
export function isFeatureSupported(feature: SDKFeature, sdkVersion: string): boolean {
  const featureVersion = features[feature];
  if (!featureVersion) {
    throw new Error(
      `Invalid SDKFeature, the following versions are supported: ${Object.keys(features)}`
    );
  }
  return semver.gte(sdkVersion, featureVersion);
}

/**
 * @internal
 */
function standardizePeerDependencies(peerDependencies: any): SnackDependencyVersions {
  if (!peerDependencies) {
    return peerDependencies;
  }
  let result: SnackDependencyVersions = peerDependencies;
  for (const name in peerDependencies) {
    const peerDep = peerDependencies[name];
    if (typeof peerDep === 'string' || peerDep === null) {
      // :thumbsup: regular peer-dependency
    } else if (typeof peerDep === 'object' && typeof peerDep.version === 'string') {
      result = result === peerDependencies ? { ...peerDependencies } : result;
      result[name] = peerDep.version;
    } else {
      // Invalid peer-dependency
      result = result === peerDependencies ? { ...peerDependencies } : result;
      delete result[name];
    }
  }

  return result;
}

/**
 * Converts older dependency formats into the SnackDependencies type.
 */
export function standardizeDependencies(dependencies: any): SnackDependencies {
  let result: SnackDependencies = dependencies;
  for (const name in dependencies) {
    const dep = dependencies[name];
    if (typeof dep === 'string') {
      result = result === dependencies ? { ...dependencies } : result;
      result[name] = {
        version: dep,
      };
    } else if (typeof dep === 'object') {
      const peerDependencies = standardizePeerDependencies(dep.peerDependencies);
      if (dep.peerDependencies !== peerDependencies) {
        result = result === dependencies ? { ...dependencies } : result;
        result[name] = {
          ...dep,
          peerDependencies,
        };
      }
    } else {
      // Invalid dependency
      result = result === dependencies ? { ...dependencies } : result;
      delete result[name];
    }
  }
  return result;
}
