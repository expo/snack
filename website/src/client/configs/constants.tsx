export default {
  links: {
    itunes: 'https://itunes.apple.com/app/apple-store/id982107779?pt=17102800&amp;ct=www&amp;mt=8',
    playstore: 'https://play.google.com/store/apps/details?id=host.exp.exponent',
    authorDocs: 'https://github.com/expo/snack/blob/main/docs/embedding-snacks.md',
  },
  appetize: {
    url: 'https://appetize.io',
    public_keys: {
      main: {
        android: 'xc1w6f1krd589zhp22a0mgftyw',
        ios: '8bnmakzrptf1hv9dq7v7bnteem',
      },
      secondary: {
        android: 'af5yxyj38991wyb09c5p7n1790',
        ios: 'tq08t4qj50qjtmgg716hn6jut4',
      },
    },
  },
  preview: {
    minWidth: 700,
  },
};
