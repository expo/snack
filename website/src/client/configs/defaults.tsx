import type { SnackFiles, SnackDependencies } from '../types';

export const DEFAULT_METADATA_NAME = 'Snack - React Native in the browser';
export const DEFAULT_METADATA_DESCRIPTION_EMPTY = `Write code in Expo's online editor and instantly use it on your phone.`;
export const DEFAULT_METADATA_DESCRIPTION_SAVED = `Try this project on your phone! Use Expo's online editor to make changes and save your own copy.`;

export const DEFAULT_DESCRIPTION = 'No description';

// TODO(cedric): Drop `import React from 'react';` when dropping SDK 45
export const DEFAULT_CODE: SnackFiles = {
  'App.js': {
    contents: `import * as React from 'react';
import { Text, View, StyleSheet } from 'react-native';
import Constants from 'expo-constants';

// You can import from local files
import AssetExample from './components/AssetExample';

// or any pure javascript modules available in npm
import { Card } from 'react-native-paper';

export default function App() {
  return (
    <View style={styles.container}>
      <Text style={styles.paragraph}>
        Change code in the editor and watch it change on your phone! Save to get a shareable url.
      </Text>
      <Card>
        <AssetExample />
      </Card>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingTop: Constants.statusBarHeight,
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
`,
    type: 'CODE',
  },
  'assets/snack-icon.png': {
    contents:
      'https://snack-code-uploads.s3.us-west-1.amazonaws.com/~asset/2f7d32b1787708aba49b3586082d327b',
    type: 'ASSET',
  },
  'components/AssetExample.js': {
    contents: `import * as React from 'react';
import { Text, View, StyleSheet, Image } from 'react-native';

export default function AssetExample() {
  return (
    <View style={styles.container}>
      <Text style={styles.paragraph}>
        Local files and assets can be imported by dragging and dropping them into the editor
      </Text>
      <Image style={styles.logo} source={require('../assets/snack-icon.png')} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    padding: 24,
  },
  paragraph: {
    margin: 24,
    marginTop: 0,
    fontSize: 14,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  logo: {
    height: 128,
    width: 128,
  }
});
`,
    type: 'CODE',
  },
  'README.md': {
    contents: `# Sample Snack app

Open the \`App.js\` file to start writing some code. You can preview the changes directly on your phone or tablet by scanning the **QR code** or use the iOS or Android emulators. When you're done, click **Save** and share the link!

When you're ready to see everything that Expo provides (or if you want to use your own editor) you can **Download** your project and use it with [expo-cli](https://docs.expo.dev/get-started/installation/#expo-cli)).

All projects created in Snack are publicly available, so you can easily share the link to this project via link, or embed it on a web page with the \`<>\` button.

If you're having problems, you can tweet to us [@expo](https://twitter.com/expo) or ask in our [forums](https://forums.expo.dev/c/expo-dev-tools/61) or [Discord](https://chat.expo.dev/).

Snack is Open Source. You can find the code on the [GitHub repo](https://github.com/expo/snack).
`,
    type: 'CODE',
  },
};

export const DEFAULT_DEPENDENCIES: SnackDependencies = {
  'react-native-paper': {
    version: '4.9.2',
    // The handle ensures that the dependency doesn't need to be resolved
    // on startup
    handle: 'snackager-1/react-native-paper@4.9.2',
  },
  '@expo/vector-icons': {
    version: '*',
  },
  'expo-constants': {
    version: '*',
  },
};
