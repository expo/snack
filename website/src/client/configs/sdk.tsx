/**
 * Setting the value to `false` will hide the version from the version picker.
 * It can be selected by specifying `?sdkVersion=99.0.0` in the URL. Ex: http://snack.expo.io?sdkVersion=99.0.0
 * This is useful when deploying the website  with preliminary support for a new SDK version.
 */
export const versions: { [version: string]: boolean } = {
  '38.0.0': true,
  '39.0.0': true,
  '40.0.0': true,
  '41.0.0': true,
};

export const DEFAULT_SDK_VERSION = '41.0.0';
export const TEST_SDK_VERSION = '38.0.0';
