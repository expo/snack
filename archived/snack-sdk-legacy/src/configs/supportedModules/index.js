import { type SDKVersion } from '../sdkVersions';

const supportedVersions: { [key: SDKVersion]: { [key: string]: string } | undefined } = {
  // No longer supported
};

export default supportedVersions;
