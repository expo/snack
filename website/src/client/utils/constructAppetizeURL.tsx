import querystring from 'query-string';

import constants from '../configs/constants';

type Props = {
  experienceURL: string;
  platform: 'android' | 'ios';
  previewQueue: 'main' | 'secondary';
  autoplay?: boolean;
  screenOnly?: boolean;
  scale?: number;
  payerCode?: string;
  deviceColor?: 'black' | 'white';
};

export default function constructAppetizeURL({
  experienceURL,
  platform,
  screenOnly = false,
  scale,
  autoplay,
  payerCode,
  previewQueue,
  deviceColor = 'black',
}: Props) {
  const appetizeOptions = {
    screenOnly,
    scale,
    autoplay: !!autoplay,
    embed: true,
    device: platform === 'ios' ? 'iphone12' : 'pixel4',
    launchUrl: platform === 'android' ? experienceURL : undefined,
    xdocMsg: true,
    deviceColor,
    xDocMsg: true,
    orientation: 'portrait',
    debug: true,
    pc: payerCode,
  };

  const appetizeKey = constants.appetize.public_keys[previewQueue][platform];
  const appParams = {
    EXDevMenuDisableAutoLaunch: true,
    EXKernelLaunchUrlDefaultsKey: experienceURL,
    EXKernelDisableNuxDefaultsKey: true,
  };

  return `${constants.appetize.url}/embed/${appetizeKey}?${querystring.stringify(
    appetizeOptions
  )}&params=${encodeURIComponent(JSON.stringify(appParams))}`;
}
