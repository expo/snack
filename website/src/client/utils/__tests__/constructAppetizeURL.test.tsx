import { TEST_SDK_VERSION } from '../../configs/sdk';
import constructAppetizeURL from '../constructAppetizeURL';

it('constructs appetize URL', () => {
  expect(
    constructAppetizeURL({
      experienceURL: `exp://exp.host/@snack/sdk.${TEST_SDK_VERSION}-456a768`,
      platform: 'ios',
      previewQueue: 'main',
    })
  ).toMatchSnapshot();
  expect(
    constructAppetizeURL({
      experienceURL: `exp://exp.host/@snack/sdk.${TEST_SDK_VERSION}-456a768`,
      platform: 'android',
      autoplay: true,
      screenOnly: true,
      scale: 2,
      payerCode: 'asdf',
      previewQueue: 'secondary',
    })
  ).toMatchSnapshot();
});
