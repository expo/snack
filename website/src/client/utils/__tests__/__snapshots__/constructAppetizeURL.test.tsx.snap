// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructs appetize URL 1`] = `"https://appetize.io/embed/8bnmakzrptf1hv9dq7v7bnteem?autoplay=false&debug=true&device=iphone12&deviceColor=black&embed=true&orientation=portrait&screenOnly=false&xDocMsg=true&xdocMsg=true&params=%7B%22EXDevMenuDisableAutoLaunch%22%3Atrue%2C%22EXKernelLaunchUrlDefaultsKey%22%3A%22exp%3A%2F%2Fexp.host%2F%40snack%2Fsdk.39.0.0-456a768%22%2C%22EXKernelDisableNuxDefaultsKey%22%3Atrue%7D"`;

exports[`constructs appetize URL 2`] = `"https://appetize.io/embed/af5yxyj38991wyb09c5p7n1790?autoplay=true&debug=true&device=pixel4&deviceColor=black&embed=true&launchUrl=exp%3A%2F%2Fexp.host%2F%40snack%2Fsdk.39.0.0-456a768&orientation=portrait&pc=asdf&scale=2&screenOnly=true&xDocMsg=true&xdocMsg=true&params=%7B%22EXDevMenuDisableAutoLaunch%22%3Atrue%2C%22EXKernelLaunchUrlDefaultsKey%22%3A%22exp%3A%2F%2Fexp.host%2F%40snack%2Fsdk.39.0.0-456a768%22%2C%22EXKernelDisableNuxDefaultsKey%22%3Atrue%7D"`;
